### PR TITLE
`Quiz exercises`: Short Answer Number Spot

### PR DESCRIPTION
**Pull request by rriyaldhi**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Server
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I implemented the changes with a good performance and prevented too many database calls.
- [x] I documented the Java code using JavaDoc style.
#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/) and ensured that the layout is responsive.
- [x] Following the [theming guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-tests/).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context

[#5459](https://github.com/ls1intum/Artemis/issues/5459)

### Description

- Added type to Short Answer Spot
- Modified Short Answer Submission
- Added Add Number Spot button in Short Answer Question configuration
- Added Validation of Number Spot Solution
- Modified the input type of Short Answer Spot during Quiz Exercise

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Student
- 1 Course

1. Log in to Artemis as an Instructor
2. Navigate to Course Management
3. Select the Course
4. Create a Quiz Exercise
5. Add Short Answer Question
6. Clear the Question Template
7. Click **Add Number Spot**
8. The option of number spot is in inclusive range format (e.g. `[0.5|10]` to set the solution between 0.5 and 10 inclusive)
9. Input other text that contains a number (e.g. `Test 1.23456`)
10. Click Visual Tab
11. Highlight the text (i.e. `Test`) with the cursor
12. Click **Add Number Spot** (This should not create a new spot because the highlighted text is not a number)
13. Highlight the number (i.e. `1.23456`) with the cursor
14. Click **Add Number Spot** (This will create a new number spot whose answer is the highlighted number `[1.23456|1.23456]`)
15. Save the Quiz Exercise
16. Click **Preview Solution** (Verify that the solutions are displayed correctly)
17. Start the Quiz Exercise
18. Log out and Log in as a Student
19. Start the Quiz Exercise
20. Input to the number spot should be limited to a number only
21. Submit (Verify that the grade is calculated correctly)

#### Code Review
- [ ] Review 1
- [ ] Review 2
#### Manual Tests
- [ ] Test 1
- [ ] Test 2

### Test Coverage

| Class/File | Branch | Line |
|------------|-------:|-----:|
| ShortAnswerSubmittedText.java | 87% | 95% |
| ShortAnswerSpot.java | 75% | 98% |
| quiz-exercise-detail.component.ts | 80.28% | 95.62% |
| quiz-exercise-validation.directive.ts | 91.01% | 97.33 |
| short-answer-question-edit.component.ts | 73.21% | 94.24% |
| short-answer-question.component.ts | 80.59% | 92.55% |
| short-answer-question-util.service.ts | 81.35% | 95.58% |

### Screenshots

Screen Shot 2022-11-24 at 2 12 35 PM
Screen Shot 2022-11-24 at 2 13 28 PM
Screen Shot 2022-11-24 at 2 14 19 PM

### CODE REVIEW

File: src/main/webapp/app/exercises/quiz/manage/short-answer-question/short-answer-question-edit.component.ts [hunk 1]
```diff
@@ -467,6 +484,15 @@
         this.questionUpdated.emit();
     }
 
+    private markdownSpotForHtml(htmlContent: string, type: SpotType): string {
+        const markdown = markdownForHtml(htmlContent);
+        if (type === SpotType.NUMBER) {
+            // If htmlContent is a number (e.g. 1.234567), markdownForHtml will escape the dot (e.g. 1\\.234567).
+            return markdown.replace('\\', ''); // Unescape the dot
```
[the remark this answers is not in the document]
Comment (rriyaldhi) [reply src/main/webapp/app/exercises/quiz/manage/short-answer-question/short-answer-question-edit.component.ts]: Changed the 1st parameter of replace to regex.

File: src/main/java/de/tum/in/www1/artemis/domain/quiz/ShortAnswerSubmittedText.java [hunk 2]
```diff
@@ -93,14 +97,41 @@ public boolean isSubmittedTextCorrect(String submittedText, String solution) {
...
+            if (matcher.find()) {
+                // Solution is in range format
+                double lowerBound = Double.parseDouble(matcher.group(1));
+                double upperBound = Double.parseDouble(matcher.group(4));
+                return submitted >= lowerBound && submitted <= upperBound;
+            }
+            else {
+                // Solution is a single number
+                double solutionInt = Double.parseDouble(solution.trim());
+                return submitted == solutionInt;
```
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/domain/quiz/ShortAnswerSubmittedText.java:122]: The solution does not have to be an int as the variable name would suggest

File: src/main/webapp/app/exercises/quiz/manage/short-answer-question/short-answer-question-edit.component.ts [hunk 3]
```diff
@@ -269,12 +271,16 @@ export class ShortAnswerQuestionEditComponent implements OnInit, OnChanges, Afte
...
-            .split(/\[-spot/g)
-            .map((splitText) => splitText.split(/\]/g))
-            .slice(1)
-            .map((sliceText) => sliceText[0]);
+        // Split into spots to generate this structure: [{spotID: "1", spotType: "text"}, {spotID: "2", spotType: "text"}, {spotID: "3", spotType: "number"}]
+        const spotParts: { spotID: string; spotType: string }[] = [];
+        const regexp = new RegExp('\\[-spot-?(number)? (\\d+)\\]', 'g');
+        let match = regexp.exec(questionText);
+        while (match !== null) {
+            const spotType = match[1] ? match[1] : 'text';
```
Comment (JohannesStoehr) [inline src/main/webapp/app/exercises/quiz/manage/short-answer-question/short-answer-question-edit.component.ts:279]:
```suggestion
            const spotType = match[1] ?? 'text';
```

File: src/main/webapp/app/exercises/quiz/manage/short-answer-question/short-answer-question-edit.component.ts [hunk 4]
```diff
@@ -269,12 +271,16 @@ export class ShortAnswerQuestionEditComponent implements OnInit, OnChanges, Afte
...
-        const spotParts = questionText
-            .split(/\[-spot/g)
-            .map((splitText) => splitText.split(/\]/g))
-            .slice(1)
-            .map((sliceText) => sliceText[0]);
+        // Split into spots to generate this structure: [{spotID: "1", spotType: "text"}, {spotID: "2", spotType: "text"}, {spotID: "3", spotType: "number"}]
+        const spotParts: { spotID: string; spotType: string }[] = [];
+        const regexp = new RegExp('\\[-spot-?(number)? (\\d+)\\]', 'g');
+        let match = regexp.exec(questionText);
+        while (match !== null) {
```
Comment (JohannesStoehr) [inline src/main/webapp/app/exercises/quiz/manage/short-answer-question/short-answer-question-edit.component.ts:278]:
```suggestion
        while (match) {
```

File: src/main/webapp/app/exercises/quiz/manage/short-answer-question/short-answer-question-edit.component.ts [hunk 5]
```diff
@@ -343,13 +350,14 @@ export class ShortAnswerQuestionEditComponent implements OnInit, OnChanges, Afte
...
+     * @param type Type of the spot that will be added
      * @desc Add the markdown for a spot at the current cursor location and
      * an option connected to the spot below the last visible row
      */
-    addSpotAtCursor(): void {
+    addSpotAtCursor(type: SpotType): void {
         const editor = this.questionEditor.getEditor();
         const optionText = editor.getCopyText();
-        const addedText = '[-spot ' + this.numberOfSpot + ']';
+        const addedText = `[-spot${type === SpotType.NUMBER ? '-number' : ''} ` + this.numberOfSpot + ']';
```
Comment (JohannesStoehr) [inline src/main/webapp/app/exercises/quiz/manage/short-answer-question/short-answer-question-edit.component.ts:360]: Why do you not use a format string for the whole line?

File: src/main/webapp/app/exercises/quiz/manage/short-answer-question/short-answer-question-edit.component.ts [hunk 6]
```diff
@@ -450,7 +463,11 @@ export class ShortAnswerQuestionEditComponent implements OnInit, OnChanges, Afte
         this.textParts = this.shortAnswerQuestionUtil.divideQuestionTextIntoTextParts(questionText);
         const textOfSelectedRow = this.textParts[selectedTextRowColumn[0]][selectedTextRowColumn[1]];
         this.textParts[selectedTextRowColumn[0]][selectedTextRowColumn[1]] =
-            textOfSelectedRow.substring(0, startOfRange) + '[-spot ' + this.numberOfSpot + ']' + textOfSelectedRow.substring(endOfRange);
+            textOfSelectedRow.substring(0, startOfRange) +
+            `[-spot${type === SpotType.NUMBER ? '-number' : ''} ` +
+            this.numberOfSpot +
+            ']' +
```
Comment (JohannesStoehr) [inline src/main/webapp/app/exercises/quiz/manage/short-answer-question/short-answer-question-edit.component.ts:469]: Why not use a format string for the whole line?

File: src/main/webapp/app/exercises/quiz/manage/short-answer-question/short-answer-question-edit.component.ts [hunk 7]
```diff
@@ -709,7 +740,9 @@ export class ShortAnswerQuestionEditComponent implements OnInit, OnChanges, Afte
 
         this.textParts = this.parseQuestionTextIntoTextBlocks(this.shortAnswerQuestion.text!);
 
-        this.textParts = this.textParts.map((part) => part.filter((text) => !text || !text.includes('[-spot ' + spotToDelete.spotNr + ']')));
+        this.textParts = this.textParts.map((part) =>
+            part.filter((text) => !text || !text.includes(`[-spot${spotToDelete.type === SpotType.NUMBER ? '-number' : ''} ` + spotToDelete.spotNr + ']')),
```
Comment (JohannesStoehr) [inline src/main/webapp/app/exercises/quiz/manage/short-answer-question/short-answer-question-edit.component.ts:744]: Why not use the format string for the whole string?

File: src/main/webapp/i18n/de/quizExercise.json [hunk 8]
```diff
@@ -184,6 +184,7 @@
                 "misleadingCorrectMapping": "Frage {{index}}: Es gibt mindestens eine korrekte Zuordnung (oder Kombination an korrekten Zuordnungen), die wenn sie verwendet wird, keine korrekte Gesamtlösung mehr zulässt.",
                 "shortAnswerQuestionEverySpotHasASolution": "Frage {{index}}: Es gibt mindestens ein Eingabefeld, dass keine Lösung besitzt.",
                 "shortAnswerQuestionEveryMappedSolutionHasASpot": "Frage {{index}}: Es fehlt mindestens ein Eingabefeld, um eine Kombination an korrekten Zuordnungen zu erhalten.",
+                "shortAnswerQuestionEverySpotNumberHasAValidSolution": "Frage {{index}}: Es gibt mindestens ein Eingabefeld, das hat ungültige Lösung.",
```
Comment (JohannesStoehr) [inline src/main/webapp/i18n/de/quizExercise.json:187]:
```suggestion
                "shortAnswerQuestionEverySpotNumberHasAValidSolution": "Frage {{index}}: Es gibt mindestens ein Eingabefeld mit einer ungültigen Lösung.",
```

File: src/test/cypress/fixtures/quiz_exercise_fixtures/shortAnswerQuiz_template.json [hunk 9]
```diff
@@ -6,7 +6,7 @@
     "matchLetterCase": false,
     "similarityValue": 85,
     "title": "Cypress SA",
-    "text": "Never gonna [-spot 1] you up\nNever gonna [-spot 2] you down\nNever gonna [-spot 3] around and [-spot 4] you\nNever gonna make you [-spot 5]\nNever gonna say [-spot 6]\nNever gonna tell a lie and hurt you",
+    "text": "Never gonna [-spot 1] you up\nNever gonna [-spot 2] you down\nNever gonna [-spot 3] around and [-spot 4] you\nNever gonna make you [-spot 5]\nNever gonna say [-spot 6]\nNever gonna tell a lie and hurt you\n[-spot-number 7], [-spot-number 8], [-spot-number 9], and [-spot-number 10]",
```
Comment (JohannesStoehr) [inline src/test/cypress/fixtures/quiz_exercise_fixtures/shortAnswerQuiz_template.json:9]: I like that text 🤣

**Review by JohannesStoehr — CHANGES_REQUESTED** [review]
Left some comments

File: src/main/webapp/app/exercises/quiz/manage/short-answer-question/short-answer-question-edit.component.ts (hunk 5 above)
Comment (rriyaldhi) [reply src/main/webapp/app/exercises/quiz/manage/short-answer-question/short-answer-question-edit.component.ts]: I changed only the part that I needed. But I've just updated the whole line to use string interpolation instead.

File: src/main/webapp/app/exercises/quiz/manage/short-answer-question/short-answer-question-edit.component.ts (hunk 6 above)
Comment (rriyaldhi) [reply src/main/webapp/app/exercises/quiz/manage/short-answer-question/short-answer-question-edit.component.ts]: I changed only the part that I needed. But I've just updated the whole line to use string interpolation instead.

**Review by LiaLism — CHANGES_REQUESTED** [review]
Highlighting a number (step 13 in testing) did not work. When I try to highlight the number in "Visual mode" it drags the square with the number, instead of highlighting the number. It worked when not switching to the visual tab. Everything else worked.

File: src/main/java/de/tum/in/www1/artemis/config/Constants.java [hunk 10]
```diff
@@ -257,6 +257,18 @@ public final class Constants {
...
+     * The regex matches range of decimal numbers
+     * For example: -10.05--1.5
+     * Group 1: -10.05 (Lower Bound)
+     * Group 2: 10.05
+     * Group 3: .05
+     * Group 4: -1.5 (Upper Bound)
+     * Group 5: 1.5
+     * Group 6: .5
+     */
+    public static final String SHORT_ANSWER_NUMBER_SPOT_TYPE_SOLUTION_REGEX = "^(-?(\\d+(\\.\\d+)?|\\.\\d+))\\s*-\\s*(-?(\\d+(\\.\\d+)?|\\.\\d+))$";
```
Comment (stefanwaldhauser) [inline src/main/java/de/tum/in/www1/artemis/config/Constants.java:270]: I would not make it that complicated tbh. Just always let the user input a range in a format that is easily parsable without that complicated regex. And always make it inclusive in both ends: For example:
- Only three is allowed `[3|3]`
- Only 4.3 is allowed `[4.3|4.3]`
- 2 to 5.7 is allowed `[2|5.7]`



Simple then remove the `[]` from both ends of the string and split by `|`.

**Review by stefanwaldhauser — CHANGES_REQUESTED** [review]
I personally find the regex approach to complicated

**Review by kekskatze — DISMISSED** [review]
Tested on ts2. Everything seemed to work well

**Review by timo07k — APPROVED** [review]
Everything works as expected apart from one extreme edge case: one can enter one "e" into a number spot, and after the submission it is gone !) Tested on TS1.

**Review by PhilipWerz — CHANGES_REQUESTED** [review]
Everything worked fine. But I could still put an "e" in the fiel where only numbers should be allowed. When I submitted the Quiz the fields where I put the e + number where empty (compare screenshot). 

Note: 
The "e" could be allowed because it's sometimes used to represent exponents.

image

**Review by evgeny1984 — CHANGES_REQUESTED** [review]
As a student, I could add answers with undesired input, including "-" chars. See the image.
image

**Comment by rriyaldhi** [comment]
> Highlighting a number (step 13 in testing) did not work. When I try to highlight the number in "Visual mode" it drags the square with the number, instead of highlighting the number. It worked when not switching to the visual tab. Everything else worked.

This is fixed now.

File: src/main/java/de/tum/in/www1/artemis/config/Constants.java (hunk 10 above)
Comment (rriyaldhi) [reply src/main/java/de/tum/in/www1/artemis/config/Constants.java]: I adjusted the implementation based on your suggestion. Thanks!

**Comment by rriyaldhi** [comment]
> Everything works as expected apart from one extreme **edge case**: one can enter one "e" into a number spot, and after the submission it is gone!) Tested on TS1.

This is fixed now. Invalid input will be cleared upon focusing out of the field.

**Comment by rriyaldhi** [comment]
> Everything worked fine. But I could still put an "e" in the fiel where only numbers should be allowed. When I submitted the Quiz the fields where I put the e + number where empty (compare screenshot).
> 
> Note: The "e" could be allowed because it's sometimes used to represent exponents.
> 
> image

This is fixed now. Invalid input will be cleared upon focusing out of the field.

**Comment by rriyaldhi** [comment]
> As a student, I could add answers with undesired input, including "-" chars. See the image. image Tested on the ts2, the rest looks awesome and works as expected!

This is fixed now. Invalid input will be cleared upon focusing out of the field.

File: src/main/webapp/i18n/de/shortAnswerQuestion.json [hunk 11]
```diff
@@ -22,7 +22,8 @@
             "editor": {
                 "addHint": "[hint] Hinweis",
                 "addExplanation": "[exp] Erklärung",
-                "addSpot": "[-spot #] Spot erzeugen",
+                "addTextSpot": "[-spot #] Text Spot erzeugen",
+                "addNumberSpot": "[-spot #] Nummer Spot erzeugen",
```
Comment (frederik-kt) [inline src/main/webapp/i18n/de/shortAnswerQuestion.json:26]:
```suggestion
                "addNumberSpot": "[-spot #] Spot für Zahl erzeugen",
```

File: src/main/webapp/i18n/de/shortAnswerQuestion.json [hunk 12]
```diff
@@ -22,7 +22,8 @@
             "editor": {
                 "addHint": "[hint] Hinweis",
                 "addExplanation": "[exp] Erklärung",
-                "addSpot": "[-spot #] Spot erzeugen",
+                "addTextSpot": "[-spot #] Text Spot erzeugen",
```
Comment (frederik-kt) [inline src/main/webapp/i18n/de/shortAnswerQuestion.json:25]:
```suggestion
                "addTextSpot": "[-spot #] Spot für Text erzeugen",
```

**Review by evgeny1984 — CHANGES_REQUESTED** [review]
If I delete the existing number spots and then add new spots, then the number continue increasing. E.g. I created 2 number spots, deleted them and then add 2 new, they will have numbers 3 and 4.
image
Also, if I follow the testing steps, I can select the string and add number spot, but it should not be possible.

**Review by kekskatze — CHANGES_REQUESTED** [review]
I created two Short-Answer questions in one quiz. On the second one adding a number spot in the visual tab didn't work. 
image

**Review by akesfeden — COMMENTED** [review]
Tested on TS2 during testing session, number input is enforced while participating the quiz and "Preview Solution" displays solutions correctly for the instructor. I didn't notice any issue related to the PR apart from the ones reported above.

File: src/main/webapp/i18n/de/shortAnswerQuestion.json (hunk 12 above)
Comment (rriyaldhi) [reply src/main/webapp/i18n/de/shortAnswerQuestion.json]: Thanks!

**Review by LiaLism — APPROVED** [review]
Adding short answer number spots worked as expected. Tested on ts1.

**Comment by evgeny1984** [comment]
When I delete the options created previously, the counter continues to increase the number. This issue is still not resolved.
image

**Review by timo07k — CHANGES_REQUESTED** [review]
If you delete an existing number spot and then add a new one, the number increases (as @evgeny1984 also mentioned). Tested on TS1.

**Pull request closed without merging**